KAFKA-13973: Fix inflated block cache metrics

nicktelford opened the pull request:
All block cache metrics are being multiplied by the total number of column families. In a `RocksDBTimestampedStore`, we have 2 column families (the default, and the timestamped values), which causes all block cache metrics in these stores to become doubled.

The cause is that our metrics recorder uses `getAggregatedLongProperty` to fetch block cache metrics. `getAggregatedLongProperty` queries the property on each column family in the database, and sums the results.

Since we always configure all column families to share the same block cache, that causes the same block cache to be queried multiple times for its metrics, with the results added togehter, effectively multiplying the real value by the total number of column families.

To fix this, we should simply use `getLongProperty`, which queries a single column family (the default one). Since all column families share the same block cache, querying just one of them will give us the correct metrics for that shared block cache.

Note: the same block cache is shared among all column families of a store irrespective of whether the user has configured a shared block cache across multiple stores.

nicktelford committed:
KAFKA-13973: Fix inflated block cache metrics

All block cache metrics are being multiplied by the total number of
column families. In a `RocksDBTimestampedStore`, we have 2 column
families (the default, and the timestamped values), which causes all
block cache metrics in these stores to become doubled.

The cause is that our metrics recorder uses `getAggregatedLongProperty`
to fetch block cache metrics. `getAggregatedLongProperty` queries the
property on each column family in the database, and sums the results.

Since we always configure all column families to share the same block
cache, that causes the same block cache to be queried multiple times for
its metrics, with the results added togehter, effectively multiplying
the real value by the total number of column families.

To fix this, we should simply use `getLongProperty`, which queries a
single column family (the default one). Since all column families share
the same block cache, querying just one of them will give us the correct
metrics for that shared block cache.

Note: the same block cache is shared among all column families of a store
irrespective of whether the user has configured a shared block cache
across multiple stores.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorder.java:
```diff
@@ -373,14 +373,14 @@ private Gauge<BigInteger> gaugeToComputeBlockCacheMetrics(final String propertyN
                         // BigInteger and construct the object from the byte representation of the value
                         result = new BigInteger(1, longToBytes(
-                            valueProvider.db.getAggregatedLongProperty(ROCKSDB_PROPERTIES_PREFIX + propertyName)
+                            valueProvider.db.getLongProperty(ROCKSDB_PROPERTIES_PREFIX + propertyName)
```
This indeed seems a bug! Thanks for finding and fixing it!

cadonna commented on streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorder.java:
```diff
@@ -373,14 +373,14 @@ private Gauge<BigInteger> gaugeToComputeBlockCacheMetrics(final String propertyN
                         // BigInteger and construct the object from the byte representation of the value
                         result = result.add(new BigInteger(1, longToBytes(
-                            valueProvider.db.getAggregatedLongProperty(ROCKSDB_PROPERTIES_PREFIX + propertyName)
+                            valueProvider.db.getLongProperty(ROCKSDB_PROPERTIES_PREFIX + propertyName)
```
How do you know that column families share the same block cache?

cadonna replied:
In this [blog post](https://www.speedb.io/blog-posts/how-does-rocksdb-memory-management-work), it says that: 

> Each column family can have its own block cache but may also be shared between column families or multiple database instances

We need to understand how we can discover both cases and compute the metric accordingly.

cadonna replied:
I think you are right about that we share the same cache across the two column families. We create and add a cache to an options object and pass the same option object to both column families. I am waiting for confirmation from the speedb chat.

nicktelford replied:
Yeah, so block caches are configurable per-CF in both RocksDB and Speedb, but the specific implementation in `TimestampedRocksDBStore` always assigns the same block cache to both column families. In fact, the design of `RocksDBStore` makes it difficult to assign different block caches to column families.

The only way an implementation could have different block caches for different column families is if `RocksDBStore` was extended and overridden, such that the configuration of the `TableFormatConfig` is provided per-CF. In which case, such a custom implementation would necessitate a custom `RocksDBMetricsRecorder`, as it's tightly coupled to `RocksDBStore` anyway.

cadonna replied:
I got confirmation, the column families share the cache.

mjsax replied:
Btw: Maybe https://github.com/speedb-io/speedb/issues/583#issuecomment-1616710762 helps?

cadonna commented:
@nicktelford Could you please add new unit tests for the case of multiple column families and adapt existing unit tests to your fix?

nicktelford commented:
> @nicktelford Could you please add new unit tests for the case of multiple column families and adapt existing unit tests to your fix?

No existing tests appear to be affected by this bug/change, but I can most likely add a test that verifies the bug and fix.

cadonna commented:
I think this part of the unit tests should be affected since the called method changes:
https://github.com/apache/kafka/blob/fd6c9f16bacf26261fee2755c1d3679b9703a8a0/streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderGaugesTest.java#L234

nicktelford committed:
Update tests for new behaviour

Our tests were verifying the use of `getAggregatedLongProperty`, when
they should instead use `getLongProperty` for _only_ block cache
metrics.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderGaugesTest.java:
```diff
@@ -211,14 +215,19 @@ private void runAndVerifySumOfProperties(final String propertyName) throws Excep
+        if (aggregated) {
+            when(dbToAdd1.getAggregatedLongProperty(ROCKSDB_PROPERTIES_PREFIX + propertyName)).thenReturn(recordedValue1);
+            when(dbToAdd2.getAggregatedLongProperty(ROCKSDB_PROPERTIES_PREFIX + propertyName)).thenReturn(recordedValue2);
+        } else {
+            when(dbToAdd1.getLongProperty(ROCKSDB_PROPERTIES_PREFIX + propertyName)).thenReturn(recordedValue1);
+            when(dbToAdd2.getLongProperty(ROCKSDB_PROPERTIES_PREFIX + propertyName)).thenReturn(recordedValue2);
+        }
```
I would move this part into `runAndVerifyBlockCacheMetricsWithMultipleCaches()`. Before we forwarded the call to this because the verifications were basically the same. That changed now, so I think it is cleaner to separate both.

nicktelford replied:
Done.

cadonna reviewed (commented):
@nicktelford Thanks for the update!

Could you also look why we did not catch this bug in `RocksDBMetricsIntegrationTest`or other metrics integration tests and add tests if needed?

nicktelford committed:
Simplify tests

nicktelford commented:
> @nicktelford Thanks for the update!
> 
> Could you also look why we did not catch this bug in `RocksDBMetricsIntegrationTest`or other metrics integration tests and add tests if needed?

@cadonna Looks like that test doesn't verify the value of those metrics; it only checks the number of metrics registered under each name (i.e. that the expected metrics are registered and available, but not what they are).

cadonna commented:
@nicktelford Yes, you are right about the existing tests! Sorry I was in a hurry and only skimmed them and missed that fact.
I would like to have some tests that verify the behavior with one single column family (non-timestamped RocksDB) and multiple column families (timestamped RocksDB) so that we get a signal when something changes on RocksDB side and to kind of document the behavior.

nicktelford committed:
Add RocksDBBlockCacheMetricsTest

Tests that `RocksDBStore` and `RocksDBTimestampedStore` produce the
expected values in their block cache metrics.

The test has been verified to fail without the fix provided by #14317,
and pass with the fix applied.

Note: the primary constructor of `RocksDBTimestampedStore` was made
`public` to match the visibility of the same constructor on the parent
`RocksDBStore`.

nicktelford commented:
@cadonna I've added a test now that I think tests this suitably. Crucially, the test fails without this fix applied, and passes with this fix applied, as you would expect.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBBlockCacheMetricsTest.java:
```diff
@@ -0,0 +1,133 @@
+import org.apache.kafka.streams.state.internals.RocksDBTimestampedStore;
+import org.apache.kafka.test.MockInternalProcessorContext;
+import org.apache.kafka.test.TestUtils;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
```
We are migrating to JUnit 5. New tests should use JUnit 5. Could you please adapt the test?

nicktelford replied:
I've pushed this change, although it's not particularly pretty: JUnit 5 doesn't have support for Parameterized tests where the test parameters are shared among all tests in the class. This issue is tracked here: https://github.com/junit-team/junit5/issues/944

I suspect this might be an issue for various other tests in the codebase as they get migrated to JUnit 5.

For now, we have to setup and teardown the state store on every test method, which slows it down a bit.

cadonna reviewed (commented):
Thanks for the updates, @nicktelford !

I have one ask. 

For the rest the PR looks good to me!

nicktelford committed:
Use JUnit 5 for tests

cadonna reviewed (approved):
LGTM!

@nicktelford Thanks again!

mjsax commented:
Cherry-picked this to `3.6` branch. Also tries to cherry-pick to `3.5` but the newly added test fails on 3.5 -- not sure why yet.
